runtime/v2/bridge.go

```diff
@@ -0,0 +1,217 @@
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package v2
+
+import (
+	"context"
+	"fmt"
+
+	"google.golang.org/grpc"
+	"google.golang.org/protobuf/types/known/emptypb"
+
+	v2 "github.com/containerd/containerd/api/runtime/task/v2"
+	v3 "github.com/containerd/containerd/api/runtime/task/v3"
+	"github.com/containerd/ttrpc"
+)
+
+// NewTaskClient returns a new task client interface which handles both GRPC and TTRPC servers depending on the
+// client object type passed in.
+//
+// Supported client types are:
+// - *ttrpc.Client
+// - grpc.ClientConnInterface
+//
+// In 1.7 we support TaskService v2 (for backward compatibility with existing shims) and GRPC TaskService v3.
+// In 2.0 we'll switch to TaskService v3 only for both TTRPC and GRPC, which will remove overhead of mapping v2 structs to v3 structs.
+func NewTaskClient(client interface{}) (v2.TaskService, error) {
+	switch c := client.(type) {
+	case *ttrpc.Client:
+		return v2.NewTaskClient(c), nil
+	case grpc.ClientConnInterface:
+		return &grpcBridge{v3.NewTaskClient(c)}, nil
+	default:
+		return nil, fmt.Errorf("unsupported shim client type %T", c)
+	}
+}
+
+// grpcBridge implements `v2.TaskService` interface for GRPC shim server.
+type grpcBridge struct {
+	client v3.TaskClient
+}
+
+var _ v2.TaskService = (*grpcBridge)(nil)
+
+func (g *grpcBridge) State(ctx context.Context, request *v2.StateRequest) (*v2.StateResponse, error) {
+	resp, err := g.client.State(ctx, &v3.StateRequest{
+		ID:     request.GetID(),
+		ExecID: request.GetExecID(),
+	})
```

Comment:
Minor nit after looking again, but for all of these that don't return an emptypb we should probably just return nil for the actual responses to be idiomatic

Reply:
Once we release 1.7 and we switch to v3, we can remove this code (as there will be no need to convert between v2 and v3 contracts), so it'll be a lot slimmer (like [sandbox/bridge.go](https://github.com/containerd/containerd/blob/8ef298d86394613085d43549c2f7b2df6fb2ce88/sandbox/bridge.go))

Reply:
Fine with me then